### ISSUE TITLE
Use rebase instead of merge for pulling in changes into fork branch?

### ISSUE THREAD

**Issue by abhijitvalluri** [fork-maintainers/iceraven-browser#92]
Should we use "squash and merge" instead of a simple merge? Or at the very least a rebase instead of merges?

I personally prefer a squash and merge (all the changes from one PR get squashed). For pulling in upstream changes from mozilla, a rebase will be nice. The advantage of this, in my opinion, is that the history of the repository is cleaner. There is one clear line of commits rather than multiple lines merging and splitting etc making it hard to follow the history. 

Also, rebasing makes things like reverting older commits far more straightforward as there is no branches to worry about. We do it this way at work as well and in general it is a much easier way to deal with things. I know this is a hotly debated topic, but if we all agree with the same approach, at least we can make a decision.

**Comment by ghost** [comment]
Yeah man do it

**Comment by interfect** [comment]
We can definitely squash and merge or rebase new features onto our main branch.

~Rebasing our commits onto later Mozilla commits would involve resolving conflicts in the distant past on our line of development that were introduced in new Mozilla commits, right? So if we then wanted to build to test to make sure we resolved the conflicts properly (and the automated merge didn't introduce bugs), we'd have to build and test what would seem to be older versions of our codebase. I'm not sure doing all that work just so any of our commits could exist on top of the latest of their commits is worth it.~ That's not what you meant at all.

Rebasing Mozilla's commits onto our commits might be more feasible, because we don't have to spelunk through very old versions of anything, and our version ends up with the clean history. It looks like they rebase PRs onto their main line, so if we merge (by merge) a PR against Mozilla Firefox and then later pull in their latest code, I think we'll already end up with two copies of the commit (original and Mozilla-rebased), right? So doing rebases ourselves won't cause more trouble.

I'm generally a merge fan, because I find it hard to follow the actual flow of what was developed on top of what, and what people were thinking, when everything is rebased. I also find it natural to "pull in" changes across a very deep split, like between us and Mozilla, so you have merges that bridge the gap at more recent points.

I just experimented with rebasing Mozilla's commits onto ours manually, and I ran into some trouble (preserved for posterity in https://github.com/fork-maintainers/iceweasel/tree/upstream-sync-3).

I did:
```
git fetch origin
git fetch upstream
git checkout upstream/master
git log
# At this point I decided I wanted to rebase an old commit from their history, and then the rest on top of the result of that, to simulate adopting this practice and doing it multiple times as Mozilla produces more commits.
git checkout 87bd44145fa3fa439eb74e32c0c6ec39ba309fdb
git branch -D upstream-sync
git checkout -b upstream-sync
git rebase -i origin/fork
# This went fine. Then I tried to take everything else
git checkout upstream/master
git checkout -b upstream-sync-2
git rebase -i upstream-sync
# Git is smart enough to not try and do all the commits again, just the ones that are new.
# But it got stuck at 806f74abe because it was empty after conflict resolution. I did as it suggested:
git commit --allow-empty
# And then made it continue:
git rebase --continue
# And similarly when it got stuck at b0729f655:
git commit --allow-empty
git rebase --continue
# Then I decided to try rebasing the same thing I just rebased, to see how it works when there's nothing to do
git checkout upstream/master
git checkout -b upstream-sync-3
git rebase -i upstream-sync-2
# It thinks those two commits that became empty still need to be rebased! And once again it gets stuck and I have to:
git commit --allow-empty
git rebase --continue
git commit --allow-empty
git rebase --continue
# Now I take a look at my final history:
git log
```

At the end I ended up with three copies of "For #14680 - Revert ObserverRegistry changes to WifiConnectionMonitor" and "Fix #14950 - Make the "Add new tab" button visible in private mode on light theme (#14951)" in the history. As far as I can tell, I'll get more copies every time I rebase Mozilla's history onto this branch, forever, if I keep handling the empty commits this way.

I suspect your new tab button visibility fix is empty because you made the same change completely separately on both branches. I'm not sure what's weird about the ObserverRegistry changes revert commit.

Did I screw this up by using `git commit --allow-empty` as git suggested, when I should have just immediately done `git rebase --continue`? Maybe git forgot I was rebasing when it made the suggestion that I run `git commit --allow-empty`. Usually when resolving conflicts during a rebase you let `git rebase --continue` create the commit.

Here's the whole (misleading?) message from Git 2.17.1:

```
$ git rebase -i upstream-sync
The previous cherry-pick is now empty, possibly due to conflict resolution.
If you wish to commit it anyway, use:

    git commit --allow-empty

Otherwise, please use 'git reset'
interactive rebase in progress; onto a59aee5d4
Last command done (1 command done):
   pick 806f74abe For #14680 - Revert ObserverRegistry changes to WifiConnectionMonitor
Next commands to do (4 remaining commands):
   pick b0729f655 Fix #14950 - Make the "Add new tab" button visible in private mode on light theme (#14951)
   pick 8e7d10c33 Import l10n.
You are currently rebasing branch 'upstream-sync-2' on 'a59aee5d4'.

Untracked files:
	env.sh
	fork.txt
	s
	testlog.txt

nothing added to commit but untracked files present
Could not apply 806f74abe... For #14680 - Revert ObserverRegistry changes to WifiConnectionMonitor
```

(This is why I prefer merges: fewer failure modes. Merges as far as I can tell can only result in conflicts, things that aren't conflicts but should have been, and refusals to merge unrelated histories. Here with the rebases, within a few commands I'm exploring [pretty barren search results, including one that's a commit to Git on Github suggesting I may have hit a Git bug](https://duckduckgo.com/?q="git+rebase"+"The+previous+cherry-pick+is+now+empty").)

Can you show how to make this work @abhijitvalluri? Do we want to just always let Github do it?

**Comment by abhijitvalluri** [comment]
Oof, that sounds rough. Let me see if I can do it any better. In theory, I thought it should work nicely, but perhaps it is more complex than that. Obviously rebasing is trivial if the branches are like so:
<p align="center">
<img src="https://user-images.githubusercontent.com/17151327/92901903-c6355d80-f418-11ea-9433-d70e4c27601b.png" width="600" />
</p>

Of course, with master renamed to our "fork" and feature branch being "upstream/master". But given the state in which the repo is right now, it is probably more challenging. Let me see if I figure out how to do a rebase easily, otherwise we can stick with merges if that is less hassle.

**Comment by abhijitvalluri** [comment]
I have a PR to bring in the changes from mozilla via rebase #94. 

Reading your above comment, I am not sure why you would want to rebase in that manner for doing this going into the future.  I think the workflow should simply be the following steps afresh everytime we want to pull in changes:
```bash
$ git fetch mozilla # I named mozilla's remote as mozilla
$ git checkout mozilla/master
$ git checkout -b mozilla_main # Create a branch so that we are not on a detached head
$ git rebase fork
```

Then, we create a PR with the `mozilla_main` or whatever we call the branch that has the latest mozilla commits on top of our fork. We create a PR to merge this into fork. In the PR we can simply do a merge (because I believe that should do a simple FF merge which will be the same as doing rebase right?) or also a rebase as it should be the same. Then our history will be clean. And then from the latest commit in fork we can repeat the same process as above for the next batch of commits. I do not see any issues doing this repeatedly. How about you?

**Comment by abhijitvalluri** [comment]
Hmm, I guess I can see why this may become a problem. Because, when we repeatedly rebase mozilla/master into our origin/fork, then both fork and mozilla master will forever have the same exact point where they diverge right? 

So, then perhaps into the future, when we keep rebasing, git will try to replay all the commits since the divergence in the history every single time? Is that the case? In which case, in the future we may end up with the "empty commits" issue you mentioned?

**Comment by abhijitvalluri** [comment]
Looks like there was an issue rebasing. So, while rebasing itself was fine and led to no errors, after I merged my PR https://github.com/fork-maintainers/iceweasel/pull/94 and checked on the main page of this repo, github was still claiming that our fork was 22 commits behind mozilla/master, despite merging those commits into our fork So, it seems like github is not capable of correctly detecting that the rebase has brought our fork inline with mozilla/master which can be confusing. So, I decided to revert all that and simply do a merge.

Merge was also weird though, for me. Locally I was able to merge without conflicts. But when I did the same on github, it was saying there is a conflict. So, I had to merge locally first and then create a PR for that. And merge that PR again, leading to two merge commits. Oh well, at least that worked. 

I am closing this because rebase seems to be more hassle than it is worth. Let us continue with merging.
